### PR TITLE
Support torch_dtype modification and update FAQs for AWQ quantization

### PR DESCRIPTION
**Pull request by AllentDan**

### CODE REVIEW

File: lmdeploy/lite/quantization/awq.py [hunk 1]
```diff
@@ -43,8 +43,10 @@
     'MixtralDecoderLayer': {
         'input_layernorm':
         ['self_attn.k_proj', 'self_attn.q_proj', 'self_attn.v_proj'],
-        'post_attention_layernorm':
-        ['block_sparse_moe.experts.{i}.w1', 'block_sparse_moe.experts.{i}.w3']
+        'post_attention_layernorm': [
+            'block_sparse_moe.gate', 'block_sparse_moe.experts.{i}.w1',
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:47]: Should we quantize `moe.gate`?

File: docs/en/quantization/w4a16.md [hunk 2]
```diff
@@ -128,3 +128,8 @@ We benchmarked the Llama-2-7B-chat and Llama-2-13B-chat models with 4-bit quanti
 | ---------------- | ------- | ------- | --------- |
 | Llama-2-7B-chat  | 112.9   | 159.4   | 206.4     |
 | Llama-2-13B-chat | N/A     | 90.7    | 115.8     |
+
+## FAQs
+
+1. Out of Memory error during quantization due to insufficient GPU memory: This can be addressed by reducing the parameter `--calib-seqlen`, increasing the parameter `--calib-samples`, and set `--batch-size` to 1.
+2. NaN value error during quantization: This may be due to a large number of zeros in the weights. For a solution, refer to the modifications in https://github.com/InternLM/lmdeploy/pull/1844/files.
```
Comment (lvhan028) [inline docs/en/quantization/w4a16.md:135]: I think we should address the NaN issue. Giving users an option to clamp the value is a solution as long as the root cause is massive zeros in the weights

File: lmdeploy/lite/quantization/awq.py (hunk 1 above)
Comment (AllentDan) [reply lmdeploy/lite/quantization/awq.py:47]: It is skipped during quantization. This is for weight migration.

File: lmdeploy/lite/apis/auto_awq.py [hunk 3]
```diff
@@ -56,6 +55,8 @@ def auto_awq(model: str,
              search_scale: bool = False,
              device: str = 'cuda',
              revision: str = None,
+             dtype: Literal['float16', 'bfloat16', 'auto'] = 'auto',
+             clamp_zeros: bool = False,
```
Comment (lvhan028) [inline lmdeploy/lite/apis/auto_awq.py:59]: propose `clamp_zero`

File: lmdeploy/lite/quantization/awq.py [hunk 4]
```diff
@@ -116,11 +118,15 @@ def skipped_module(name: str):
...
+def get_weight_scale(weight, q_group_size=-1, clamp_zeros: bool = False):
     org_shape = weight.shape
     if q_group_size > 0:
         weight = weight.view(-1, q_group_size)
-    scale = weight.abs() / weight.abs().amax(dim=1, keepdim=True)
+    if clamp_zeros:
+        scale = weight.abs() / weight.abs().amax(dim=1,
+                                                 keepdim=True).clamp(min=1e-4)
+    else:
+        scale = weight.abs() / weight.abs().amax(dim=1, keepdim=True)
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:129]: Can we check if `weight.abs().amax(dim=1, keepdim=True)` equals 0?
If it is zero, can we clamp zero to a small value by default? If we can, may provide warning to users, informing it might cause performance degradation.

File: lmdeploy/lite/quantization/awq.py [hunk 5]
```diff
@@ -116,11 +118,15 @@ def skipped_module(name: str):
...
 @torch.no_grad()
-def get_weight_scale(weight, q_group_size=-1):
+def get_weight_scale(weight, q_group_size=-1, clamp_zeros: bool = False):
     org_shape = weight.shape
     if q_group_size > 0:
         weight = weight.view(-1, q_group_size)
-    scale = weight.abs() / weight.abs().amax(dim=1, keepdim=True)
+    if clamp_zeros:
+        scale = weight.abs() / weight.abs().amax(dim=1,
+                                                 keepdim=True).clamp(min=1e-4)
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:127]: why choose `1e-4`? Is it robust?

File: lmdeploy/lite/quantization/awq.py (hunk 4 above)
Comment (lvhan028) [reply lmdeploy/lite/quantization/awq.py:129]: My another concern is the justifiability of `--clamp-zero` option
Because it is clear that the divisor cannot be zero.

File: lmdeploy/lite/quantization/awq.py [hunk 6]
```diff
@@ -228,10 +246,10 @@ def smooth_fc_fcs(pre_fc: torch.nn.Module,
         fc.weight.mul_(scales.view(1, -1))
 
     for p in pre_fc.parameters():
-        assert torch.isnan(p).sum() == 0
+        assert torch.isnan(p).sum() == 0, 'may use `--clamp-zeros` option'
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:249]: Is the 0 divisor when calculating scale the only reason to NAN?

File: lmdeploy/lite/utils/load.py [hunk 7]
```diff
@@ -21,10 +23,14 @@ def load_hf_from_pretrained(pretrained_model_name_or_path,
                                            trust_remote_code=True)
 
     # HACK hard code for qwen, other configs do not have the `fp16` attribute.
-    if dtype == torch.float16:
-        hf_config.fp16 = True
-    elif dtype == torch.bfloat16:
-        hf_config.bf16 = True
+    if hasattr(hf_config, 'fp16') or hasattr(hf_config, 'bf16'):
```
Comment (lvhan028) [inline lmdeploy/lite/utils/load.py:26]: qwen1 or qwen2?

Comment (AllentDan) [reply lmdeploy/lite/utils/load.py:26]: Only qwen1. These are compatible codes.

File: lmdeploy/lite/quantization/awq.py (hunk 6 above)
Comment (AllentDan) [reply lmdeploy/lite/quantization/awq.py]: Yes, the only reason.

File: lmdeploy/lite/quantization/awq.py (hunk 5 above)
Comment (AllentDan) [reply lmdeploy/lite/quantization/awq.py]: Just the clamp value awq used.

File: docs/en/quantization/w4a16.md [hunk 8]
```diff
@@ -128,3 +128,7 @@ We benchmarked the Llama-2-7B-chat and Llama-2-13B-chat models with 4-bit quanti
 | ---------------- | ------- | ------- | --------- |
 | Llama-2-7B-chat  | 112.9   | 159.4   | 206.4     |
 | Llama-2-13B-chat | N/A     | 90.7    | 115.8     |
+
+## FAQs
+
+1. Out of Memory error during quantization due to insufficient GPU memory: This can be addressed by reducing the parameter `--calib-seqlen`, increasing the parameter `--calib-samples`, and set `--batch-size` to 1.
```
Comment (lvhan028) [inline docs/en/quantization/w4a16.md:134]: 缺失了 "2. 量化时，无法链接huggingface并下载数据集。可以尝试使用镜像，`export HF_ENDPOINT=https://hf-mirror.com`。"

File: lmdeploy/lite/apis/auto_awq.py [hunk 9]
```diff
@@ -56,6 +55,7 @@ def auto_awq(model: str,
              search_scale: bool = False,
              device: str = 'cuda',
              revision: str = None,
+             dtype: Literal['float16', 'bfloat16', 'auto'] = 'auto',
```
Comment (lvhan028) [inline lmdeploy/lite/apis/auto_awq.py:58]: miss the docstring

File: lmdeploy/lite/apis/calibrate.py [hunk 10]
```diff
@@ -205,6 +205,7 @@ def calibrate(model: str,
               w_bits: int = 4,
               w_group_size: int = 128,
               search_scale: bool = False,
+              dtype: str = 'auto',
```
Comment (lvhan028) [inline lmdeploy/lite/apis/calibrate.py:208]: Let's keep `dtype` declaration consistent with it in `auto_awq`

File: lmdeploy/lite/apis/gptq.py [hunk 11]
```diff
@@ -29,9 +29,6 @@ def auto_gptq(model: str,
         calib_seqlen (int): The sequence length for calibration.
         w_bits (int): Bit number for weight quantization.
         w_group_size (int): Group size for weight quantization statistics.
-        search_scale (bool): Whether search scale ratio. Default to False,
```
Comment (lvhan028) [inline lmdeploy/lite/apis/gptq.py:32]: Should we add `dtype` for `auto_gptq`?

File: lmdeploy/lite/apis/smooth_quant.py [hunk 12]
```diff
@@ -19,8 +20,8 @@ def smooth_quant(model: str,
                  search_scale: bool = False,
                  batch_size: int = 1,
                  w_bits: int = 8,
+                 dtype: str = 'auto',
```
Comment (lvhan028) [inline lmdeploy/lite/apis/smooth_quant.py:23]: Let's keep dtype declaration consistent with it in auto_awq

File: docs/en/quantization/w4a16.md (hunk 8 above)
Comment (AllentDan) [reply docs/en/quantization/w4a16.md:134]: 英文用户应该没这问题？

File: lmdeploy/lite/quantization/awq.py [hunk 13]
```diff
@@ -153,8 +160,12 @@ def smooth_ln_fcs(ln: torch.nn.Module,
     concat_w = torch.cat([fc.weight for fc in fcs], dim=0)
     w_scales = get_weight_scale(concat_w, group_size)
 
+    w_scales_pow = w_scales.pow(1 - alpha)
+    if w_scales_pow.min().item() == 0:
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:164]: w_scales are clamped.
It's impossible that `w_scales.pow(1-alpha) == 0`

File: lmdeploy/lite/quantization/calibration.py [hunk 14]
```diff
@@ -291,8 +294,12 @@ def _search_module_scale(block, linears2scale: list, x, kwargs={}):
         org_sd = {k: v.cpu() for k, v in block.state_dict().items()}
         for ratio in range(0, n_grid):
             ratio = ratio * 1 / n_grid
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/calibration.py:296]: no need " * 1"

File: lmdeploy/lite/quantization/calibration.py [hunk 15]
```diff
@@ -291,8 +294,12 @@ def _search_module_scale(block, linears2scale: list, x, kwargs={}):
...
-            scales = (x_max.pow(ratio) /
-                      w_mean.pow(1 - ratio)).clamp(min=1e-4).view(-1)
+            w_mean_pow = w_mean.pow(1 - ratio)
+            if w_mean_pow.min().item() == 0:
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/calibration.py:298]: Given `w_mean >= 1e-4` and `ratio in [0, 1/20, 2/20, ..., 19/20]`, w_mean_pow > 0.00015
Therefore this branch is always False. It's unnecessary to clamp it here

File: lmdeploy/lite/quantization/awq.py (hunk 13 above)
Comment (AllentDan) [reply lmdeploy/lite/quantization/awq.py:164]: Mathematically, it is impossible, but from previous experience, it is possible.

File: lmdeploy/lite/quantization/awq.py [hunk 16]
```diff
@@ -204,8 +215,12 @@ def smooth_fc_fcs(pre_fc: torch.nn.Module,
...
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:219]: This branch is always False. The reason is provided in `calibration.py`

File: lmdeploy/lite/quantization/awq.py (hunk 13 above)
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:164]: This branch is always False. The reason is provided in calibration.py

File: lmdeploy/lite/quantization/calibration.py (hunk 15 above)
Comment (AllentDan) [reply lmdeploy/lite/quantization/calibration.py:298]: w_mean contains zeros.

File: lmdeploy/lite/quantization/awq.py [hunk 17]
```diff
@@ -120,7 +122,12 @@ def get_weight_scale(weight, q_group_size=-1):
     org_shape = weight.shape
     if q_group_size > 0:
         weight = weight.view(-1, q_group_size)
-    scale = weight.abs() / weight.abs().amax(dim=1, keepdim=True)
+    abs_weight = weight.abs()
+    abs_weight_amax = abs_weight.amax(dim=1, keepdim=True)
+    if abs_weight_amax.min().item() == 0:
+        print('clamp zero values to a minimal decimal')
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:128]: print('weight.amax.min is zero, clamping weight.amax to 1e-4')

**Review by lvhan028 — APPROVED** [review]

**Comment by lvhan028** [comment]
test passed with model `InternVL2_5-78B`

**Review by RunningLeon — COMMENTED** [review]
may merge the latest main to fix ut

**Review by RunningLeon — COMMENTED** [review]
tested ok on Qwen-VL-Chat

**Review by RunningLeon — APPROVED** [review]
LGTM
